### PR TITLE
Add missing tests for a DateBox widget

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

**Review by DokaRus — DISMISSED** [review]

File: testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js [hunk 1]
```diff
@@ -1131,6 +1131,40 @@ module('Advanced caret', setupModule, () => {
         assert.deepEqual(this.keyboard.caret(), { start: 3, end: 5 }, 'caret was moved');
     });
 
+    test('Move caret to the next group if advanceCaret is changed at runtime', function(assert) {
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js]: The `advanceCaret` option is:
- not documented
- always `true`
- never used

Let's just remove it.

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 2]
```diff
@@ -2797,6 +2816,46 @@ QUnit.module('datebox with time component', {
         }
     });
 
+    QUnit.test('date box should chenge behavior if adaptivityEnabled option is changed at runtime', function(assert) {
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]:
```suggestion
    QUnit.test('date box should change behavior if adaptivityEnabled option is changed at runtime', function(assert) {
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 3]
```diff
@@ -2797,6 +2816,46 @@ QUnit.module('datebox with time component', {
...
+        const originalWidthFunction = renderer.fn.width;
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: You can use stub's restore method instead:
```

const widthStub = sinon.stub(renderer.fn, 'width');
try {
...
} finally {
    widthStub.restore();
}
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js (hunk 2 above)
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: I suggest splitting this test scenario into 2 tests:
- Change `adaptivityEnabled` from `false` to `true`
- Change `adaptivityEnabled` from `true` to `false`

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 4]
```diff
@@ -2818,6 +2877,32 @@ QUnit.module('datebox with time component', {
         assert.equal($clock.length, 0, 'clock was not rendered');
     });
 
+    QUnit.test('date box should chenge behavior if showAnalogClock option is changed at runtime', function(assert) {
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]:
```suggestion
    QUnit.test('date box should change behavior if showAnalogClock option is changed at runtime', function(assert) {
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 5]
```diff
@@ -4205,6 +4308,24 @@ QUnit.module('datebox validation', {}, () => {
...
+            value: new Date(2017, 5, 5)
+        });
+        const dateBox = $dateBox.dxDateBox('instance');
+        const $input = $dateBox.find(`.${TEXTEDITOR_INPUT_CLASS}`);
+
+        $($input).trigger('change');
+        dateBox.option('dateOutOfRangeMessage', 'another test message');
+        $($input).trigger('change');
+
+        const validationError = $dateBox.dxDateBox('instance').option('validationError').message;
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]:
```suggestion
        const validationError = dateBox.option('validationError.message');
```

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 6]
```diff
@@ -2818,6 +2877,32 @@ QUnit.module('datebox with time component', {
...
+    QUnit.test('date box should chenge behavior if showAnalogClock option is changed at runtime', function(assert) {
+        const $element = $('#dateBox').dxDateBox({
+            type: 'datetime',
+            pickerType: 'calendar',
+            showAnalogClock: false,
+            opened: true
+        });
+        const instance = $element.dxDateBox('instance');
+
+        let $content = $(instance._popup.$content());
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: Looks like we already have such test:
https://github.com/DevExpress/DevExtreme/blob/b56479ce50edf5f3423dea097f64e3e86ee0beae/testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js#L2800-L2819

**Review by San4es — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js (hunk 6 above)
Comment (alexander-kotov-dx) [reply testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: Here we check option changing on initialization and change the option value from false to true at runtime

File: testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js (hunk 1 above)
Comment (alexander-kotov-dx) [reply testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js]: I going to remove it by separate PR

**Review by DokaRus — DISMISSED** [review]

File: testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js (hunk 1 above)
Comment (San4es) [reply testing/tests/DevExpress.ui.widgets.editors/datebox.mask.tests.js]: Ok. But do we need adding a test for this option that will be removed in the next PR?

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js (hunk 6 above)
Comment (San4es) [reply testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: What about reworking tests these 2 tests to a few separate ones:
- check rendering when showAnalogClock = false initially
- check rendering when showAnalogClock = true initially
- check changing showAnalogClock from false to true
- check changing showAnalogClock from false to true

?

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 7]
```diff
@@ -2797,6 +2816,60 @@ QUnit.module('datebox with time component', {
         }
     });
 
+    QUnit.test('date box should change behavior if adaptivityEnabled option is changed to false at runtime', function(assert) {
```
Comment (San4es) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js]: Parameterize?
image

**Review by DokaRus — DISMISSED** [review]

**Review by San4es — APPROVED** [review]
LGTM

**Review by DokaRus — APPROVED** [review]

**Pull request merged**